src/frontend/lsp/ocamlmerlin_lsp.ml

```diff
@@ -529,13 +532,63 @@ let on_request :
         find [] (index - 1)
     in
 
-    Document_store.get store uri >>= fun doc ->
-    let prefix = prefix_of_position (Document.source doc) position in
-    log ~title:"debug" "completion prefix: |%s|" prefix;
-    let command =
-      Query_protocol.Complete_prefix (
-        prefix,
-        position,
+    let range_prefix prefix =
+      let start_ =
+        let len = String.length prefix in
+        let character = lsp_position.character - len in
+        { lsp_position with character }
+      in
+      { Lsp.Protocol.
+        start_;
+        end_ = lsp_position;
+      }
+    in
+
+    let item index entry =
+      let prefix, (entry: Query_protocol.Compl.entry) =
+        match entry with
+        | `Keep entry -> `Keep, entry
+        | `Replace (range, entry) -> `Replace range, entry
+      in
+      let kind: Lsp.Protocol.Completion.completionItemKind option =
+        match entry.kind with
+        | `Value -> Some Value
+        | `Constructor -> Some Constructor
+        | `Variant -> None
+        | `Label -> Some Property
+        | `Module |`Modtype -> Some Module
+        | `Type -> Some TypeParameter
+        | `MethodCall -> Some Method
+      in
+      let textEdit =
+        match prefix with
+        | `Keep -> None
+        | `Replace range -> Some {
+          Lsp.Protocol.TextEdit.
+          range;
+          newText = entry.name;
+        }
+      in
+      {
+        Lsp.Protocol.Completion.
+        label = entry.name;
+        kind;
+        detail = Some entry.desc;
+        inlineDetail = None;
+        itemType = Some entry.desc;
+        documentation = Some entry.info;
```

Comment:
For the future: Might need to move `documentation` requests to `completionItem/resolve` request as including docs harms perf as discussed in https://github.com/ocaml/merlin/issues/726#issuecomment-475843295